pkg/cmd/cli/cmd/version.go

```diff
@@ -0,0 +1,184 @@
+			if err := options.Validate(args); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(cmd, err.Error()))
+			}
+
+			if err := options.Complete(out); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(cmd, err.Error()))
+			}
+
+			if err := options.RunVersion(f); err != nil {
+				kcmdutil.CheckErr(err)
+			}
+		},
+	}
+
+	return cmd
+}
+
+func (o *VersionOptions) Validate(args []string) error {
+	if len(args) > 0 {
+		return fmt.Errorf("no arguments should be passed")
+	}
+	return nil
+}
+
+func (o *VersionOptions) Complete(out io.Writer) error {
+	o.Out = out
+	o.Err = os.Stderr
+	return nil
+}
+
+// RunVersion attempts to display client and server versions for Kubernetes and OpenShift
+func (o VersionOptions) RunVersion(f *clientcmd.Factory) error {
+	fmt.Fprintf(o.Out, "%s %v\n", o.BaseName, version.Get())
+	fmt.Fprintf(o.Out, "kubernetes %v\n", kubeversion.Get())
+	if o.PrintEtcdVersion {
+		fmt.Fprintf(o.Out, "etcd %v\n", etcdversion.Version)
+	}
+
+	if o.PrintClientFeatures {
+		features := []string{}
+		if tokencmd.BasicEnabled() {
+			features = append(features, "Basic-Auth")
+		}
+		if tokencmd.GSSAPIEnabled() {
+			features = append(features, "GSSAPI")
+			features = append(features, "Kerberos") // GSSAPI or SSPI
+			features = append(features, "SPNEGO")   // GSSAPI or SSPI
+		}
+		fmt.Printf("features: %s\n", strings.Join(features, " "))
+	}
+
+	if f == nil {
+		return nil
+	}
+
+	// max amount of time we want to wait for server to respond
+	timeout := 10 * time.Second
+
+	done := make(chan error)
+	oVersion := ""
+	kVersion := ""
+	versionHost := ""
+
+	// start goroutine to fetch openshift / kubernetes server version
+	go func(f *clientcmd.Factory) {
+		defer close(done)
+
+		var err error
+		o.Client, o.KClient, err = f.Clients()
+		if err != nil {
+			done <- err
+			return
+		}
+
+		o.ClientConfig, err = f.OpenShiftClientConfig.ClientConfig()
+		if err != nil {
+			done <- err
+			return
+		}
+		versionHost = o.ClientConfig.Host
+
+		ocVersionBody, err := o.Client.Get().AbsPath("/version/openshift").Do().Raw()
+		if err != nil && !(kapierrors.IsNotFound(err) || kapierrors.IsUnauthorized(err)) {
+			done <- err
+			return
+		}
+		var ocServerInfo version.Info
+		err = json.Unmarshal(ocVersionBody, &ocServerInfo)
+		if err != nil {
+			done <- err
+			return
+		}
+		oVersion = fmt.Sprintf("%v", ocServerInfo)
+
+		kubeVersionBody, err := o.KClient.Get().AbsPath("/version").Do().Raw()
+		if err != nil && !(kapierrors.IsNotFound(err) || kapierrors.IsUnauthorized(err)) {
+			done <- err
+			return
+		}
+		var kubeServerInfo kubeversion.Info
+		err = json.Unmarshal(kubeVersionBody, &kubeServerInfo)
+		if err != nil {
+			done <- err
+			return
+		}
+		kVersion = fmt.Sprintf("%v", kubeServerInfo)
+
+	}(f)
+
+	select {
+	case err, closed := <-done:
+		if strings.HasSuffix(fmt.Sprintf("%v", err), "connection refused") {
+			return nil
+		}
+		if closed && err != nil {
+			// we want to print error instead of returning so that
```

Comment:
It works fine as long as I print to stdout, if I return the error, or print to stderr, `hack/test-cmd.sh` fails every time due to [these lines](https://github.com/openshift/origin/blob/master/hack/test-cmd.sh#L139-145):

```
Logging to /tmp/openshift/test-cmd//logs...
[WARNING] System logger could not be started, 'sar' binary not found in this environment.
openshift version:
openshift v1.3.0-alpha.0+d268f24-dirty
kubernetes v1.3.0+57fb9ac
etcd 2.3.0+git
error: Missing or incomplete configuration info.  Please login or point to an existing, complete config file:

  1. Via the command-line flag --config
  2. Via the KUBECONFIG environment variable
  3. In your home directory as ~/.kube/config

To view or setup config directly use the 'config' command.
[ERROR] PID 15431: hack/test-cmd.sh:141: `openshift version` exited with status 1.
[INFO]          Stack Trace:
[INFO]            1: hack/test-cmd.sh:141: `openshift version`
[INFO]   Exiting with code 1.
```